tests: lib: multi_heap fix up_squared test failures

laurenmurphyx64 opened the pull request:
Fixes test failures on up_squared caused by ACPI adding to heap size (K_HEAP_MEM_POOL_SIZE). The test assumed the heap size always matched the Kconfig CONFIG_HEAP_MEM_POOL_SIZE.

Fixes #79195

npitre reviewed (changes requested):
This looks fragile and wrong.

If the actual heap size is `K_HEAP_MEM_POOL_SIZE` and not necessarily
`CONFIG_HEAP_MEM_POOL_SIZE` then the former should replace the later in 
test_mheap.h.

laurenmurphyx64 commented:
> This looks fragile and wrong.
> 
> If the actual heap size is `K_HEAP_MEM_POOL_SIZE` and not necessarily `CONFIG_HEAP_MEM_POOL_SIZE` then the former should replace the later in test_mheap.h.

@npitre I tried replacing that, but the actual system heap size isn't the only problem. up_squared's ACPI (and other non-emulated board's subsystems, possibly) adds to the system heap size, and then proceeds to allocate (k_malloc) and deallocate (k_free) memory in the system heap while doing its work during startup, leaving holes and blocks scattered around. Every test case assumes that the test is the only one using the system heap, and many of them attempt to precisely "fill" the system heap (e.g. test_mheap_malloc_align4), and then check that allocating more memory fails. Trying to precisely fill the heap fails because we don't know what the ACPI put in there, and trying to check that allocating more memory fails doesn't work because it goes backwards to allocate blocks in the holes.

In particular, when I change to K_HEAP_MEM_POOL_SIZE in test_mheap.h and change SIZE in test_mheap_api.c to "#define SIZE (K_HEAP_MEM_POOL_SIZE / NMEMB / 2)" to accommodate, I'm halted by a page fault in test_mheap_malloc_align4 when it runs over the edge trying to fill the heap precisely without knowing the ACPI has filled an unknown amount of memory. (That's my guess, anyways.)

Sorry if the solution is obvious, but do you have suggestions on what I can do in this case?

npitre commented:
> Trying to precisely fill the heap fails because we don't know what the ACPI
> put in there, and trying to check that allocating more memory fails doesn't
> work because it goes backwards to allocate blocks in the holes.

Well, let's have a look:

```
        for (nb = 0; nb < ARRAY_SIZE(block); nb++) {
                /**
                 * TESTPOINT: This routine provides traditional malloc()
                 * semantics. Memory is allocated from the heap memory pool.
                 */
                block[nb] = k_malloc(BLK_SIZE_MIN);
                if (block[nb] == NULL) {
                        break;
                }
        }

        /* Some platforms have components that variably add to the mem pool
         * size and make it larger than the known CONFIG_HEAP_MEM_POOL_SIZE
         */
        if (CONFIG_HEAP_MEM_POOL_SIZE < K_HEAP_MEM_POOL_SIZE) {
                block_fail = k_malloc(K_HEAP_MEM_POOL_SIZE * 2);
        } else {
                block_fail = k_malloc(BLK_SIZE_MIN);
        }
        /** TESTPOINT: Return NULL if fail.*/
        zassert_is_null(block_fail, NULL);
```

Here you added a malloc whose size is `K_HEAP_MEM_POOL_SIZE * 2`. Of course
this is going to fail even when the heap is completely empty. That's not the
point of the test though. Here we're supposed to fill the heap with smallish
allocations until it is full, and then try one more to make sure it is still
full.

One reason why this test might fail for you could be because
`ARRAY_SIZE(block)` depends on `CONFIG_HEAP_MEM_POOL_SIZE`. If
`K_HEAP_MEM_POOL_SIZE` is bigger then you may run out of array slots
before the heap becomes full. This one and all the others but one should 
be fixed by using `K_HEAP_MEM_POOL_SIZE` in test_mheap.h as I suggested.

The one test, namely `test_mheap_realloc`, is indeed flawed. Here I'd fix it
by adding the same kind of loop as above, storing pointers from allocations
of `BLK_SIZE_MIN` into an array until the heap is really depleted before
attempting a realloc with a smaller size.

> In particular, when I change to K_HEAP_MEM_POOL_SIZE in test_mheap.h and
> change SIZE in test_mheap_api.c to "#define SIZE (K_HEAP_MEM_POOL_SIZE /
> NMEMB / 2)" to accommodate, I'm halted by a page fault in 
> test_mheap_malloc_align4 when it runs over the edge trying to fill the 
> heap precisely without knowing the ACPI has filled an unknown amount of 
> memory. (That's my guess, anyways.)

Should never cause a page fault when trying to allocate more than the heap
available space. My guess is rather `void *block[BLK_NUM_MAX]` overflowing
the stack.

laurenmurphyx64 committed:
tests: lib: multi_heap fix up_squared test failures

Fixes test failures on up_squared caused by ACPI adding to
heap size (K_HEAP_MEM_POOL_SIZE). The test assumed the heap size
always matched the Kconfig CONFIG_HEAP_MEM_POOL_SIZE=256.

Signed-off-by: Lauren Murphy <lauren.murphy@intel.com>

laurenmurphyx64 commented:
@npitre Have updated, let me know what you think

laurenmurphyx64 commented:
@kartben Do both @andyross and @npitre need to approve for this to get merged, or just one of them?

kartben commented:
> @kartben Do both @andyross and @npitre need to approve for this to get merged, or just one of them?

Only one of the assignees is fine, however due to 4.1 being in rc3 right now it's best to keep for 4.2 if that's ok. This is only updating a test but the changes certainly look non-trivial
